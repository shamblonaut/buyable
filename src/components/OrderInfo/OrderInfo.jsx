import { ShoppingBag } from "lucide-react";
import PropTypes from "prop-types";

import { Container, Title, Field, CheckoutLink } from "./OrderInfo.styles";

const OrderInfo = ({ products, cart, setCart, taxPercent }) => {
  const subtotal = products.reduce(
    (sum, product) =>
      cart[product.id] ? sum + product.price * cart[product.id] : sum,
    0,
  );
  const tax = subtotal * (taxPercent / 100);
  const total = subtotal + tax;

  const clearCart = () => {
    setCart({});
  };

  return (
    <Container>
      <Title>Order Summary</Title>
      <Field aria-label="subtotal-field">
        <label>Subtotal: </label>
        <p>${subtotal.toFixed(2)}</p>
      </Field>
      <Field aria-label="tax-field">
        <label>Tax (12%): </label>
        <p>${tax.toFixed(2)}</p>
      </Field>
      <Field aria-label="total-price-field">
        <label>Total Price: </label>
        <p>${total.toFixed(2)}</p>
      </Field>
      <CheckoutLink to="/disclaimer" onClick={clearCart}>
        <ShoppingBag /> Place Order
      </CheckoutLink>
    </Container>
  );
};

OrderInfo.propTypes = {
  products: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      price: PropTypes.number.isRequired,
    }),
  ).isRequired,
  cart: PropTypes.objectOf(PropTypes.number).isRequired,
  setCart: PropTypes.func.isRequired,
  taxPercent: PropTypes.number.isRequired,
};

export default OrderInfo;
