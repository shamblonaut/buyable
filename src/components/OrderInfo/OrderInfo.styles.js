import { Link } from "react-router-dom";
import styled from "styled-components";

export const Container = styled.div`
  border: 3px solid var(--color-dark);
  border-radius: 8px;
  padding: 0 32px;
  margin-bottom: 32px;
  padding-bottom: 0;
  width: 100%;
  max-width: 600px;

  background-color: var(--color-accent-dark);
  color: var(--color-light);

  display: flex;
  flex-direction: column;
`;

export const Title = styled.p`
  font-size: 2rem;
  font-weight: 700;
  font-family: var(--font-highlight);
  margin: 16px 0;
`;

export const Field = styled.span`
  display: flex;
  justify-content: space-between;
  margin: 8px 0;

  font-family: var(--font-highlight);
  font-size: 1.25rem;
  font-weight: 700;

  & label {
    font-family: var(--font-normal);
    margin-right: 8px;
  }
`;

export const CheckoutLink = styled(Link)`
  display: flex;
  justify-content: center;
  align-items: center;

  width: 100%;
  margin: 32px 0;
  padding: 8px 16px;
  font-size: 1.25rem;
  font-weight: 500;

  background-color: var(--color-green-dark);
  color: var(--color-light);
  border: 2px solid var(--color-dark);
  border-radius: 4px;

  transition: all 200ms ease;

  &:active {
    transform: scale(0.95);
  }

  & .lucide {
    width: 20px;
    height: auto;
    margin-right: 8px;
  }
`;
