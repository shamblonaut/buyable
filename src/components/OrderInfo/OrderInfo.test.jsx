import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, within } from "@testing-library/react";

import { BrowserRouter } from "react-router-dom";

import { mockProductList, mockCart } from "@tests/data";

import OrderInfo from "./OrderInfo";

const mockTaxPercent = 12;

const setCart = vi.fn();

describe("OrderInfo component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly", () => {
    const { container } = render(
      <BrowserRouter>
        <OrderInfo
          products={mockProductList}
          cart={mockCart}
          setCart={setCart}
          taxPercent={mockTaxPercent}
        />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it("calculates subtotal properly", () => {
    render(
      <BrowserRouter>
        <OrderInfo
          products={mockProductList}
          cart={mockCart}
          setCart={setCart}
          taxPercent={mockTaxPercent}
        />
      </BrowserRouter>,
    );

    expect(
      within(screen.getByLabelText("subtotal-field")).getByText("$3743.03"),
    ).toBeInTheDocument();
  });

  it("calculates tax properly", () => {
    render(
      <BrowserRouter>
        <OrderInfo
          products={mockProductList}
          cart={mockCart}
          setCart={setCart}
          taxPercent={mockTaxPercent}
        />
      </BrowserRouter>,
    );

    expect(
      within(screen.getByLabelText("tax-field")).getByText("$449.16"),
    ).toBeInTheDocument();
  });

  it("calculates total price properly", () => {
    render(
      <BrowserRouter>
        <OrderInfo
          products={mockProductList}
          cart={mockCart}
          setCart={setCart}
          taxPercent={mockTaxPercent}
        />
      </BrowserRouter>,
    );

    expect(
      within(screen.getByLabelText("total-price-field")).getByText("$4192.19"),
    ).toBeInTheDocument();
  });

  it("has a link to disclaimer page", () => {
    render(
      <BrowserRouter>
        <OrderInfo
          products={mockProductList}
          cart={mockCart}
          setCart={setCart}
          taxPercent={mockTaxPercent}
        />
      </BrowserRouter>,
    );

    expect(screen.getByRole("link")).toHaveAttribute("href", "/disclaimer");
  });
});
