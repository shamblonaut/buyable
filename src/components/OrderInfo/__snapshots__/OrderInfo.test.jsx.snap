// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OrderInfo component > renders correctly 1`] = `
<div>
  <div
    class="sc-bRKDuR hvBsMM"
  >
    <p
      class="sc-hvigdm eEPHTP"
    >
      Order Summary
    </p>
    <span
      aria-label="subtotal-field"
      class="sc-fhHczv irEvAa"
    >
      <label>
        Subtotal: 
      </label>
      <p>
        $
        3743.03
      </p>
    </span>
    <span
      aria-label="tax-field"
      class="sc-fhHczv irEvAa"
    >
      <label>
        Tax (12%): 
      </label>
      <p>
        $
        449.16
      </p>
    </span>
    <span
      aria-label="total-price-field"
      class="sc-fhHczv irEvAa"
    >
      <label>
        Total Price: 
      </label>
      <p>
        $
        4192.19
      </p>
    </span>
    <a
      class="sc-ggWZvA hYXIeT"
      data-discover="true"
      href="/disclaimer"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-shopping-bag"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 10a4 4 0 0 1-8 0"
        />
        <path
          d="M3.103 6.034h17.794"
        />
        <path
          d="M3.4 5.467a2 2 0 0 0-.4 1.2V20a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6.667a2 2 0 0 0-.4-1.2l-2-2.667A2 2 0 0 0 17 2H7a2 2 0 0 0-1.6.8z"
        />
      </svg>
       Place Order
    </a>
  </div>
</div>
`;
